### ISSUE TITLE
iOS example Retrained Stripped model memory usage causes app to crash without warning

### ISSUE THREAD

**Issue by mat-peterson** [tensorflow/tensorflow#2927]
### Environment info
Operating System: iOS

### Steps to reproduce
1. create graph with image_retraining example
2. use strip_unused tool on graph with input_node_names=Mul, output_node_names=final_result
3. run iOS example with the retrained stripped graph
4. 80% of the time the app crashes, apparently because it allocates too much memory

### What have you tried?
1. reducing the size of the image being processed

### Logs or other output that would be helpful

I /Users/matp/Documents/projects/mytrue/third-party/tensorflow3/tensorflow-0.9.0rc0/tensorflow/contrib/ios_examples/simple/RunModelViewController.mm:225] Session created.
I /Users/matp/Documents/projects/mytrue/third-party/tensorflow3/tensorflow-0.9.0rc0/tensorflow/contrib/ios_examples/simple/RunModelViewController.mm:228] Graph created.
I /Users/matp/Documents/projects/mytrue/third-party/tensorflow3/tensorflow-0.9.0rc0/tensorflow/contrib/ios_examples/simple/RunModelViewController.mm:235] Creating session.
W tensorflow/core/framework/op_def_util.cc:332] Op BatchNormWithGlobalNormalization is deprecated. It will cease to work in GraphDef version 9. Use tf.nn.batch_normalization().

![image](https://cloud.githubusercontent.com/assets/4634115/16151795/dae3e988-346c-11e6-86ea-8ba085165887.png)

**Comment by petewarden** [comment]
Sorry you're hitting problems! We actually have a few different ways to reduce memory usage. One of them is to run tensorflow/contrib/quantization/tools/quantize_graph with --mode=weights to shrink the size of the model on disk by quantizing the weights to eight bits. I'm also working on folding in the batch normalization ops into the weights, which will reduce memory pressure.

We also have memory-mapped constants, which you can try by running this script on your graph:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/util/convert_graphdef_memmapped_format.cc

I haven't tried this myself yet on iOS, but with previous projects it was a great way to reduce memory problems since mapped files don't seem to count towards your overall usage, and are automatically swapped out when pressure is high.

**Comment by jeffxtang** [comment]
@mat-peterson have you tried Pete's suggestion and does it work? I'm still having problem using the quantized model (same as https://github.com/tensorflow/tensorflow/issues/3619).

@petewarden i'm trying to understand why the Inception v1 can work with the iOS simple project on actual device (and Run Model many times) but a retrained and stripped model based on it simply crashes on device when Run Model? What's the difference there? Any insights would be appreciated!

**Comment by jeffxtang** [comment]
Oh I just realized one probably big difference: the inception v1 model is about 50+MB and my retrained model, which is based on inception v3, is close to 90MB. Maybe that's why. 

See my updates in https://github.com/tensorflow/tensorflow/issues/3619 on how to resolve the memory issue using the quantized v1 model running on actual device.

**Comment by petewarden** [comment]
I'm hopeful that these issues are solved with the new approach documented here:
https://petewarden.com/2016/09/27/tensorflow-for-mobile-poets/

I'm closing this bug, but please reopen if you're still seeing the same issue.

**Comment by jeffxtang** [comment]
Thanks @petewarden. For a relatively old approach but one that works for me in building an iOS app published in App Store, please take a look at my blog documenting the whole process of using a retrained, stripped and quantized inception v3 model at http://jeffxtang.github.io
